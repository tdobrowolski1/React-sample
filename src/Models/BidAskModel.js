class BidAskModel {
    Value;
    Lot;
    
}

export default BidAskModel