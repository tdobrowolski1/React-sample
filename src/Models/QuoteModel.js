class QuoteModel {
    Ask;
    Bid;
    Ticker;
}

export default QuoteModel;