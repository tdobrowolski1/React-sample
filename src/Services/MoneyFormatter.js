
class MoneyFormater {
    //mocking it for sample EURUSD scenario
    FormatBeginPrice(price) {
        return price.toString().substring(0,1) + "." +price.toString().substring(1,3) ;

    }
    FormatMidPrice(price){
        return price.toString().substring(3,5)
    }

    FormatEndPrice(price){
        return price.toString().substring(5,6) ;
    }

}

export default MoneyFormater;