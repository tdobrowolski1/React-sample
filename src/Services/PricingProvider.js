import QuoteModel from '../Models/QuoteModel'
import BidAskModel from '../Models/BidAskModel'
class PricingProvider {
    constructor(randomGenerator) {
        // could use some interfacing/ injection / Hoc (depending on pattern adopted) 
        this.randomGenerator = randomGenerator;
    }

    GetQuote(ticker) {
        let quote = new QuoteModel();
        quote.Bid = new BidAskModel()
        quote.Ask = new BidAskModel();
        quote.Ticker = ticker;
        quote.Bid.Value = this.randomGenerator.GetRandomRange(140000, 160000);
        quote.Bid.Lot = this.randomGenerator.GetRandomRange(1, 100);
        quote.Ask.Value =  this.randomGenerator.GetRandomRange(quote.Bid.Value + 1, quote.Bid.Value + 100);
        quote.Ask.Lot = this.randomGenerator.GetRandomRange(1, 100);
        return quote;
    }
}

export default PricingProvider;