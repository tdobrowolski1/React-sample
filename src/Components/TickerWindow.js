import React, { PureComponent } from 'react';
import PricingProvider from '../Services/PricingProvider';
import PropTypes from 'prop-types';
import MoneyFormatter from '../Services/MoneyFormatter';
import PriceBox from '../Components/PriceBox';
import '../App.css';
class TickerWindow extends PureComponent {
    componentDidMount = () => {
        setInterval(() => {
            this.setState({ Quote: this.props.pricingProvider.GetQuote(this.state.Ticker) })
        }, 500);
    }
    constructor(props) {
        super(props);
        this.state = { Quote: this.props.pricingProvider.GetQuote("EURUSD"), Ticker: "EURUSD", Lot: 5 };
    }
    handleOrder = (e, isAsk, quote) => {
        let message = `You ${isAsk ? 'bought' : 'sold'} ${this.state.Lot} ${quote.Ticker}`
        alert(message)
    }
    LotChange(e) {
        if (e.target.value !== "") {
            this.setState({ Lot: e.target.value });
        }

    }

    PairChange(e) {
        this.setState({Ticker : e.target.value});
    }
    render() {
        return (
            <div className="TickerWindow">
                <div className="TickerHeader">
                </div>
                <div> <select onChange={this.PairChange.bind(this)}><option value="EURUSD">EUR/USD</option><option value="GBPUSD">GBP/USD</option><option value="CHFUSD">CHF/USD</option></select></div>
                <div className="PriceBoxContainer">
                    <PriceBox isAsk={false} quote={this.state.Quote} moneyFormatter={this.props.moneyFormatter} onClickCallback={this.handleOrder} />
                </div>
                <div className="PriceBoxContainer">
                    <PriceBox isAsk={true} quote={this.state.Quote} moneyFormatter={this.props.moneyFormatter} onClickCallback={this.handleOrder} />
                </div>
                <div><input type="number" id="lot" value={this.state.Lot} onChange={this.LotChange.bind(this)} ></input></div>
            </div>
        );
    }
}
TickerWindow.propTypes = {
    moneyFormatter : PropTypes.instanceOf(MoneyFormatter),
    pricingProvider : PropTypes.instanceOf(PricingProvider)
}

export default TickerWindow;
