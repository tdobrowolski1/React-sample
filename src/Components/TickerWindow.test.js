import React from 'react';
import TickerWindow from './TickerWindow'
import renderer from 'react-test-renderer';
import MoneyFormatter from '../Services/MoneyFormatter';
import PricingProvider from '../Services/PricingProvider';
import RandomNumberGenerator from '../Services/RandomNumberGenerator'
import QuoteModel from '../Models/QuoteModel'
import BidAskModel from '../Models/BidAskModel'
import { configure } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
describe('TickerBox', () => {
    var randomNumberGenerator ;
    var pricingProvider ;
    var moneyFormatter;
    var quote;
    beforeEach(() => {
        randomNumberGenerator = new RandomNumberGenerator();
        randomNumberGenerator.GetRandomRange = jest.fn().mockReturnValue(123456);
        pricingProvider = new PricingProvider(randomNumberGenerator);
        moneyFormatter = new MoneyFormatter();
        quote = new QuoteModel();
        quote.Ticker = "EURUSD";
        quote.Bid = new BidAskModel();
        quote.Ask = new BidAskModel();
        quote.Bid.Value = 123456;
        quote.Ask.Value = 123456;
    })
    it('renders correctly', () => {
        const ticker = renderer
            .create(<TickerWindow  pricingProvider = {pricingProvider} moneyFormatter={moneyFormatter} />)
            .toJSON();
        expect(ticker).toMatchSnapshot();

    });


});