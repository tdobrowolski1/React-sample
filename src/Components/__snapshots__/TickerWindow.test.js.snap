// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickerBox renders correctly 1`] = `
<div
  className="TickerWindow"
>
  <div
    className="TickerHeader"
  />
  <div>
     
    <select
      onChange={[Function]}
    >
      <option
        value="EURUSD"
      >
        EUR/USD
      </option>
      <option
        value="GBPUSD"
      >
        GBP/USD
      </option>
      <option
        value="CHFUSD"
      >
        CHF/USD
      </option>
    </select>
  </div>
  <div
    className="PriceBoxContainer"
  >
    <div
      className="PriceBox bid"
      onClick={[Function]}
    >
      <div
        className="PriceBoxHeader floatLeft"
      >
        EURUSD
      </div>
      <div
        className="PriceBoxHeader floatRight"
      >
        1.23
      </div>
      <div
        className="PriceBoxBody"
      >
        <span
          className="mediumPrice"
        >
           
          45
        </span>
         
        <span
          className="smallPrice"
        >
          6
        </span>
      </div>
    </div>
  </div>
  <div
    className="PriceBoxContainer"
  >
    <div
      className="PriceBox ask"
      onClick={[Function]}
    >
      <div
        className="PriceBoxHeader floatRight"
      >
        EURUSD
      </div>
      <div
        className="PriceBoxHeader floatLeft"
      >
        1.23
      </div>
      <div
        className="PriceBoxBody"
      >
        <span
          className="mediumPrice"
        >
           
          45
        </span>
         
        <span
          className="smallPrice"
        >
          6
        </span>
      </div>
    </div>
  </div>
  <div>
    <input
      id="lot"
      onChange={[Function]}
      type="number"
      value={5}
    />
  </div>
</div>
`;
